File: android/app/src/main/java/io/github/vezzulab/juke/AppLog.kt
package io.github.vezzulab.juke

import android.content.Context
import android.content.res.Resources
import android.net.Uri
import android.os.Build
import android.util.Log
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * A small log file the person can read and send along with a problem report. Nothing leaves the phone by itself:
 * whatever is shown or copied goes through [redact] first (server address, login, passwords and tokens are hidden).
 */
object AppLog {
    private const val LIMIT = 256 * 1024L
    private var file: File? = null
    private val stamp = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
    private const val REPO = "https://github.com/vezzulab/juke"

    fun init(context: Context) {
        if (file != null) return
        file = File(context.filesDir, "juke.log")
        val previous = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, error ->
            e("crash", "Uncaught exception in ${thread.name}", error)
            previous?.uncaughtException(thread, error)
        }
        i("app", "---- Juke ${BuildConfig.VERSION_NAME} starting ----")
        summary().lines().forEach { i("app", it) }
    }

    @Synchronized
    private fun write(level: String, tag: String, message: String, error: Throwable?) {
        val target = file ?: return
        runCatching {
            if (target.length() > LIMIT) target.renameTo(File(target.path + ".1"))
            val trace = error?.let { StringWriter().also { w -> it.printStackTrace(PrintWriter(w)) }.toString() }.orEmpty()
            target.appendText("${stamp.format(Date())} $level $tag: $message\n${if (trace.isNotEmpty()) trace else ""}")
        }
    }

    fun i(tag: String, message: String) { Log.i(tag, message); write("INFO   ", tag, message, null) }
    fun w(tag: String, message: String, error: Throwable? = null) { Log.w(tag, message, error); write("WARNING", tag, message, error) }
    fun e(tag: String, message: String, error: Throwable? = null) { Log.e(tag, message, error); write("ERROR  ", tag, message, error) }

    fun summary(): String {
        val metrics = Resources.getSystem().displayMetrics
        return listOf(
            "Juke ${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
            "Android ${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT}) · ${Build.MANUFACTURER} ${Build.MODEL}",
            "Screen ${metrics.widthPixels}x${metrics.heightPixels} · ${metrics.densityDpi} dpi · ${Locale.getDefault().toLanguageTag()}",
        ).joinToString("\n")
    }

    private val querySecret = Regex("([?&;](?:p|t|s|u|pass|password|token|salt|apikey|api_key|key|auth)=)[^&\\s\"']+", RegexOption.IGNORE_CASE)
    private val urlLogin = Regex("(\\b[a-z][a-z0-9+.-]*://)[^/\\s:@]+:[^/\\s@]+@", RegexOption.IGNORE_CASE)

    /** Hide what should never be pasted into a public place. */
    fun redact(text: String, secrets: List<String> = emptyList()): String {
        var out = text
        secrets.filter { it.length >= 3 }.distinct().sortedByDescending { it.length }.forEach { out = out.replace(it, "<hidden>") }
        out = urlLogin.replace(out) { "${it.groupValues[1]}<hidden>@" }
        return querySecret.replace(out) { "${it.groupValues[1]}<hidden>" }
    }

    /** The newest part of the log, redacted. */
    fun read(secrets: List<String>, maxChars: Int = 60_000): String {
        val target = file ?: return ""
        val text = listOf(File(target.path + ".1"), target).filter { it.exists() }.joinToString("") { runCatching { it.readText() }.getOrDefault("") }
        val tail = if (text.length > maxChars) text.takeLast(maxChars).substringAfter('\n') else text
        return redact(tail, secrets)
    }

    fun secretsOf(server: io.github.vezzulab.juke.data.ServerConfig): List<String> =
        listOf(server.password, server.user, server.url, server.url.substringAfter("://").substringBefore('/'))

    fun issueUrl(title: String, body: String): String =
        "$REPO/issues/new?title=${Uri.encode(title)}&body=${Uri.encode(body)}"
}

File: juke/applog.py
"""A small log file, so a person can read what Juke has been doing and send it along with a problem report.

Nothing leaves the computer by itself. Whatever is shown or copied goes through ``redact`` first: your home
folder, passwords, login tokens and the address of your own server are hidden.
"""

from __future__ import annotations

import logging
import logging.handlers
import os
import platform
import re
import sys
import threading
from pathlib import Path
from typing import Iterable

from . import __version__
from .config import DATA_DIR

LOG_PATH = DATA_DIR / "juke.log"
_LIMIT = 512 * 1024                 # the file keeps about this much, plus one older copy
_log = logging.getLogger("juke")

_QUERY_SECRET = re.compile(r"(?i)([?&;](?:p|t|s|u|pass|password|token|salt|apikey|api_key|key|auth)=)[^&\s\"']+")
_URL_LOGIN = re.compile(r"(?i)(\b[a-z][a-z0-9+.-]*://)[^/\s:@]+:[^/\s@]+@")


def setup() -> None:
    """Start writing ``juke.log`` (rotating) and record crashes that would otherwise only reach a terminal."""
    try:
        DATA_DIR.mkdir(parents=True, exist_ok=True)
        handler = logging.handlers.RotatingFileHandler(LOG_PATH, maxBytes=_LIMIT, backupCount=1, encoding="utf-8")
    except OSError:
        return
    handler.setFormatter(logging.Formatter("%(asctime)s %(levelname)-7s %(name)s: %(message)s", "%Y-%m-%d %H:%M:%S"))
    root = logging.getLogger()
    if any(isinstance(h, logging.handlers.RotatingFileHandler) and getattr(h, "baseFilename", "") == str(LOG_PATH) for h in root.handlers):
        return
    root.addHandler(handler)
    root.setLevel(logging.WARNING)                  # libraries only when something is wrong (they may print addresses)
    _log.setLevel(logging.INFO)
    logging.captureWarnings(True)

    previous = sys.excepthook

    def crashed(kind, error, trace):
        _log.critical("Uncaught exception", exc_info=(kind, error, trace))
        previous(kind, error, trace)

    sys.excepthook = crashed
    inherited = threading.excepthook

    def thread_crashed(args):
        _log.critical("Uncaught exception in thread %s", getattr(args.thread, "name", "?"),
                      exc_info=(args.exc_type, args.exc_value, args.exc_traceback))
        inherited(args)

    threading.excepthook = thread_crashed
    _log.info("---- Juke %s starting ----", __version__)
    for line in system_summary().splitlines():
        _log.info(line)


def qt_messages() -> None:
    """Qt's own warnings (missing plugins, drawing problems) into the same file."""
    from PySide6.QtCore import QtMsgType, qInstallMessageHandler

    levels = {QtMsgType.QtWarningMsg: logging.WARNING, QtMsgType.QtCriticalMsg: logging.ERROR, QtMsgType.QtFatalMsg: logging.CRITICAL}
    qt_log = logging.getLogger("juke.qt")

    def handler(kind, _context, message):
        if kind in levels:
            qt_log.log(levels[kind], "%s", message)
        else:
            print(message, file=sys.stderr)

    qInstallMessageHandler(handler)


def system_summary() -> str:
    """Version and system facts that help someone reproduce a problem."""
    lines = [f"Juke {__version__}" + (" (AppImage)" if os.environ.get("APPIMAGE") else "")]
    lines.append(f"System: {_os_name()} · {platform.machine()}")
    desktop = os.environ.get("XDG_CURRENT_DESKTOP", "?")
    lines.append(f"Desktop: {desktop} · session {os.environ.get('XDG_SESSION_TYPE', '?')}")
    try:
        import PySide6
        from PySide6.QtCore import qVersion

        lines.append(f"Python {platform.python_version()} · PySide6 {PySide6.__version__} · Qt {qVersion()}")
    except Exception:
        lines.append(f"Python {platform.python_version()}")
    try:
        import vlc

        version = vlc.libvlc_get_version()
        lines.append("libVLC " + (version.decode() if isinstance(version, bytes) else str(version)).split(" ")[0])
    except Exception:
        lines.append("libVLC: not available")
    return "\n".join(lines)


def _os_name() -> str:
    try:
        for line in Path("/etc/os-release").read_text().splitlines():
            if line.startswith("PRETTY_NAME="):
                return line.split("=", 1)[1].strip().strip('"')
    except OSError:
        pass
    return platform.platform()


def redact(text: str, secrets: Iterable[str] = ()) -> str:
    """Hide what should never be pasted into a public place."""
    for secret in sorted({s for s in secrets if s and len(s) >= 3}, key=len, reverse=True):
        text = text.replace(secret, "<hidden>")
    text = _URL_LOGIN.sub(r"\1<hidden>@", text)
    text = _QUERY_SECRET.sub(r"\1<hidden>", text)
    home = str(Path.home())
    if len(home) > 1:
        text = text.replace(home, "~")
    return text


def read(secrets: Iterable[str] = (), max_bytes: int = 400_000) -> str:
    """The newest part of the log, redacted (empty when nothing has been written yet)."""
    chunks: list[str] = []
    size = 0
    for path in (Path(str(LOG_PATH) + ".1"), LOG_PATH):
        try:
            data = path.read_bytes()
        except OSError:
            continue
        chunks.append(data.decode("utf-8", "replace"))
        size += len(data)
    text = "".join(chunks)
    if len(text) > max_bytes:
        text = text[-max_bytes:]
        text = text[text.find("\n") + 1:]
    return redact(text, secrets)
